### ISSUE TITLE
Allow comments that match any previous level of indentation

### ISSUE THREAD

**Issue by tomprince** [adrienverge/yamllint#141]
Comments that match a previous level of indentation should be allowed, at the end of that level of indentation.

A minimal example that I think should lint succesfully:

```yaml
---
config:
    entry:
        - things
    # no default, so we fail on new entries
options:
    - more stuff
```
gives
```
5:5       warning  comment not indented like content  (comments-indentation)
```

**Comment by adrienverge** [comment]
This would need a new option on the `comments-indentation`. Proposals and contributions are welcome!

**Comment by itxds** [comment]
This other case also fails:
```yml
config:
  - property1: some value
    property2: some other value
  # comment
```
Also gives a `comments-indentation` warning.

**Comment by ssbarnea** [comment]
I would say this counts as a bug in indentation rule as it creates false positive like below even when using code-commenting support in your editors.

```
---
extends: default

rules:
  line-length:
    max: 999
  # document-start: disable  # <-- false positive, this must be at this level!
```

Current behavior forces people to miss-indent they temporary disable YAML code, so when they use editor to uncomment it, it will be wronly indented.

**Comment by ssbarnea** [comment]
This bug seems to become more annoying as it forces users to wrongly change indentation of commented lines, introducing bugs.

In the example below line 6 is wrongly indicated with comments-indentation when that is the
correct way to indent it. 

Most editors do have shortcuts for comment/uncomment and they do keep indentation correctly.

When yamllint runs, it gives a false positive about indentation. Most users will wrongly change indentation to make the linter happy. Big mistake, as this introduce a bug when the lines are uncommented.

This can easily pass CI and code0review, especially when lists are longer, and have disastrous effects.

I know for sure one place where current behavior invites people to mistakenly move `when:` conditions one level up or down when used inside blocks. The effect is that the condition would apple to the last task in block instead of the entire block.

```
foo:
  - foo
  - bar:
    - a
    - b
  # - zoo
```

So, please accept this proposal which is almost two years old. Having a rule about comment indentation is good and we should not be forced to disable it because is partially broken. Lets just fix it.

**Comment by adrienverge** [comment]
@ssbarnea as always, the tone of your comments is a bit annoying.

This issue is known, and in the first reply to this thread I wrote "proposals and contributions are welcome!"

> Lets just fix it.

So what are you waiting for?

PS: It's not a "bug", like you say. It's intended to work like that. If you're unhappy, please disable the rule (it's very easy and yamllint will keep working), or contribute (positively) to this project.

**Comment by ssbarnea** [comment]
@adrienverge Your last comment gave me the impression that you do not want the default behavior to be changed. If true, I doubt someone would create a PR that adds another option that is disabled by default, not when it is so easy to just disable the entire `comments-indentation` rule.

If it would be about my personal projects it would not be big deal, I already did that. The reason why I supported this change is related to scaling good practices. It is really hard to promote a linter that does invite users to break their code. I mention this because at least two different people asked me that.

What I am waiting for? I am waiting for a confirmation from the project lead regarding what kind of change would be found acceptable.

**Comment by adrienverge** [comment]
@ssbarnea you're talking about **"a bug", yamllint "inviting users to break their code", and "disastrous effects"**. Let's take a step back and stop dramatizing.

You want to change the default behavior of this rule to match your personal expectations, as well as the 2 people you've talked to (and maybe many other people). The problem is that what you call "false positives about indentation" is what some other people expect as true positives -- because your good practices are not always the other ones' good practices. Changing the default behavior **would break the rule for these users**.

Instead, we could use an opt-in configuration (proposal welcome) to achieve *your* good practice.

**Comment by scholer** [comment]
I would also vote in favor of this. We have cases where we want to temporarily comment out some parts of our yaml configuration files. But yamllint forces us to change the indentation to pass.

Consider this example (which passes yamllint):

```yaml
recipients:
  - name: John Doe
    email: john.doe@example.com
  - name: Jane Smith
    email: jane.smith@example.com
  - name: Alice Johnson
    email: alice.johnson@example.com
```

If I want to remove "Jane Smith" from the recipients list, I can just highlight the lines and comment them out like this, and yamllint accepts it just fine:

```yaml
recipients:
  - name: John Doe
    email: john.doe@example.com
  # - name: Jane Smith
  #   email: jane.smith@example.com
  - name: Alice Johnson
    email: alice.johnson@example.com

```

However, if it just so happens that we need to comment out the last entry from the list, yamllint will complain with ` error    comment not indented like content  (comments-indentation)`:

```yaml
recipients:
  - name: John Doe
    email: john.doe@example.com
  - name: Jane Smith
    email: jane.smith@example.com
  # - name: Alice Johnson
  #   email: alice.johnson@example.com

```

Now, you could say "just place the `#` at the beginning of the line!", like so, then it passes:

```yaml
recipients:
  - name: John Doe
    email: john.doe@example.com
  - name: Jane Smith
    email: jane.smith@example.com
#   - name: Alice Johnson
#     email: alice.johnson@example.com
```

The problem here is that this only works for the last entry in the list. Applying the same "place the `#` at the beginning of the line" logic fails for the other items in the list:

```yaml
recipients:
  - name: John Doe
    email: john.doe@example.com
#  - name: Jane Smith
#    email: jane.smith@example.com
  - name: Alice Johnson
    email: alice.johnson@example.com
``` 

`> 4:1       error    comment not indented like content  (comments-indentation)`

The best solution we currently have is to add an additional indentation to the commented-out items:

```yaml
recipients:
  - name: John Doe
    email: john.doe@example.com
  - name: Jane Smith
    email: jane.smith@example.com
    # - name: Alice Johnson
    #   email: alice.johnson@example.com
```

But obviously this is problematic: when the lines are uncommented again, they will have the wrong indentation.

We are considering to just disable the `comments-indentation` check entirely. However, this leaves many cases where comments are very clearly at the entirely wrong indentation level, and would surely decrease the quality and readability of our yaml files over time.

I can see that [owenlamont/ryl](https://github.com/owenlamont/ryl) now has support for this feature, so we might switch over to ryl, just to have support for this feature. But ideally we'd want to see this supported natively in yamllint.


NB: Here is our .yamllintrc configuration:

```yaml
extends: default

rules:

  # limits line length; allows long unbreakable words and inline mappings
  line-length:
    max: 150                              # maximum line length of 150 characters
    allow-non-breakable-words: true       # allow long URLs or strings without spaces
    allow-non-breakable-inline-mappings: true # extend the above rule to inline mappings.

  # prevents duplicate keys are excluded to support YAML anchors/aliases
  key-duplicates:
    forbid-duplicated-merge-keys: false   # prevents repeated merge keys used with YAML anchors and aliases

  # controls how octal literals are interpreted
  octal-values:
    forbid-implicit-octal: true   # prevents accidental octal values starting with 0 (e.g. 0777)
    forbid-explicit-octal: false  # prevents explicit octal notation (e.g. 0o755)

  # allow YAML files without explicit document start marker (---)
  document-start: disable  # ignore missing '---' at the top of YAML files

  # enforce comment indentation issues as errors instead of warnings
  comments-indentation:
    level: error  # treat indentation issues as errors

  # forbids non-explicit truthy values; only true/false are allowed (not YES, off, etc.)
  truthy: enable
```
